avalanche/training/plugins/generative_replay.py

```diff
@@ -0,0 +1,157 @@
+                 batch_size_mem: int = None,
+                 task_balanced_dataloader: bool = False,
+                 untrained_solver: bool = True):
+        '''
+        Init.
+        '''
+        super().__init__()
+        self.mem_size = mem_size
+        self.batch_size = batch_size
+        self.batch_size_mem = batch_size_mem
+        self.task_balanced_dataloader = task_balanced_dataloader
+        self.generator_strategy = generator
+        if self.generator_strategy:
+            self.generator = generator.model
+        else: 
+            self.generator = None
+        self.untrained_solver = untrained_solver
+        self.model_is_generator = False
+
+    def before_training(self, strategy: "SupervisedTemplate", *args, **kwargs):
+        """Checks whether we are using a user defined external generator 
+        or we use the strategy's model as the generator. 
+        If the generator is None after initialization 
+        we assume that strategy.model is the generator."""
+        if not self.generator_strategy:
+            self.generator_strategy = strategy
+            self.generator = strategy.model
+            self.model_is_generator = True
+
+    def before_training_exp(self, strategy: "SupervisedTemplate",
+                            num_workers: int = 0, shuffle: bool = True,
+                            **kwargs):
+        """
+        Make deep copies of generator and solver before training new experience.
+        """
+        if self.untrained_solver:
+            # The solver needs to be trained before labelling generated data and
+            # the generator needs to be trained before we can sample.
+            return
+        self.old_generator = deepcopy(self.generator)
+        self.old_generator.eval()
+        if not self.model_is_generator:
+            self.old_model = deepcopy(strategy.model)
+            self.old_model.eval()
+
+    def after_training_exp(self, strategy: "SupervisedTemplate",
+                           num_workers: int = 0, shuffle: bool = True,
+                           **kwargs):
+        """
+        Set untrained_solver boolean to False after (the first) experience,
+        in order to start training with replay data from the second experience.
+        """
+        self.untrained_solver = False
+
+    def before_training_iteration(self, strategy: "SupervisedTemplate",
+                                  **kwargs):
+        """
+        Generating and appending replay data to current minibatch before 
+        each training iteration.
+        """
+        if self.untrained_solver:
+            # The solver needs to be trained before labelling generated data and
+            # the generator needs to be trained before we can sample.
+            return
+        # extend X with replay data
+        replay = self.old_generator.generate(
+            len(strategy.mbatch[0]) * (strategy.experience.current_experience)
+            ).to(strategy.device)  
```

Comment:
I actually got inspired to do this from this [intro_to_generative_replay notebook](https://github.com/ContinualAI/colab/blob/master/notebooks/intro_to_generative_replay.ipynb) from the Avalanche/colab repo. In the literature I usually see authors employ a weighted loss instead  to achieve a similar effect (e.g. as described here in [equation 3](https://www.nature.com/articles/s41467-020-17866-2#Sec17)). It could be due to the VAE and classifier I use, but in my case the increasing generated data fared a few percentage points higher in accuracy than the weighted loss approach.